lib/ansible/plugins/lookup/fileglob.py

```diff
@@ -27,13 +27,27 @@
 
 class LookupModule(LookupBase):
 
+    GLOBS =frozenset(['?', '*', '[', ']'])
+
     def run(self, terms, variables=None, **kwargs):
 
         ret = []
         for term in terms:
-            term_file = os.path.basename(term)
-            dwimmed_path = self.find_file_in_search_path(variables, 'files', os.path.dirname(term))
-            if dwimmed_path:
-                globbed = glob.glob(to_bytes(os.path.join(dwimmed_path, term_file), errors='surrogate_or_strict'))
+
+            # find smallest unglobbed match
+            min_spot = len(term)
+            for symbol in self.GLOBS:
+                x = term.find(symbol, 0, min_spot)
+                if x > 0:
+                    min_spot = x
+
+            if min_spot == len(term):
+                dwimmed_path = self.find_file_in_search_path(variables, 'files', term)
+                ret.append(dwimmed_path)
+            else:
+                term_root = term[:min_spot]
+                dwimmed_path = self.find_file_in_search_path(variables, 'files', os.path.dirname(term_root))
+                globbed = glob.glob(to_bytes(dwimmed_path + term[min_spot:], errors='surrogate_or_strict'))
```

Comment:
The testcase playbook from the comments of issue #17136 doesn't work for me (it shows no results) because the paths are not valid. This line should probably say `globbed = glob.glob(to_bytes(os.path.join(dwimmed_path, term[min_spot:]), ...))` to properly insert separators if needed.